### PR TITLE
[BEAM-1274] Add SSL mutual authentication support

### PR DESCRIPTION
**Pull request by jbonofre**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [X] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by jbonofre** [comment]
R: @echauchot

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java [hunk 1]
```diff
@@ -257,14 +275,40 @@ public ConnectionConfiguration withPassword(String password) {
       return builder().setPassword(password).build();
     }
 
+    /**
+     * If Elasticsearch uses SSL with mutual authentication (via shield),
```
Comment (jbonofre) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java]: SSL is already a specialization of TLS.

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java [hunk 2]
```diff
@@ -285,6 +329,27 @@ public HttpAsyncClientBuilder customizeHttpClient(
...
             });
       }
+      if (getKeystorePath() != null) {
+        try {
+          KeyStore keyStore = KeyStore.getInstance("jks");
+          try (InputStream is = new FileInputStream(new File(getKeystorePath()))) {
+            keyStore.load(is, getKeystorePassword().toCharArray());
+          }
+          final SSLContext sslContext = SSLContexts.custom()
+              .loadTrustMaterial(keyStore, null).build();
```
Comment (jbonofre) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java]: I think it's better to use the keystore, and the user can add a self-signed cert in the keystore.

**Comment by jbonofre** [comment]
Agree for the itests for both user authentication and SSL mutual authentication. It just means that the elasticsearch instance has to use shield for the itests.

**Comment by echauchot** [comment]
Another comment: 
If the user provides a certificate in his keystore, is the `HttpAsynClient` going to do mutual client + server authentication?
If there is no client certificate in the keystore is it going to be server only SSL authentication?

**Comment by jbonofre** [comment]
It depends of the shield configuration:
1. if mutual authentication is needed on the ES side, then,  the client/IO has to  provide a client certificate. The client certificate has to be present and trusted in the server/ES trustsore.
2. if mutual  authentication is not enabled, then, the client/IO can provide or not the client certificate.

What do you mean by server authentication ? Username/password ?

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java (hunk 2 above)
Comment (echauchot) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java]: my bad, I miss typed the code I meant: `.loadTrustMaterial(keyStore, new TrustSelfSignedStrategy())`

**Comment by echauchot** [comment]
I meant the certificate exchange you described. I just wanted to know if the code would work in both mutual and standard SSL handshakes.

**Comment by jbonofre** [comment]
The "classic" TLS/SSL with server certificate already works and it's still working of course.

**Comment by ssisk** [comment]
I agree with the idea that SSL end to end should be tested via integration tests, not unit tests.  It would be interesting to ask if the createClient() method should be unit tested or not.  Basically - is there something we should be doing in unit tests to make sure that the setup is getting put in correctly? I'm not sure if that's a good idea or not. If there's no interesting logic, then it's probably not worth it.

cc @chamikaramj for questions about the integration tests

**Comment by jbonofre** [comment]
If needed, I can provide the Shield configuration for the Elasticsearch instance used in the integration tests.

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java [hunk 3]
```diff
@@ -257,14 +275,40 @@ public ConnectionConfiguration withPassword(String password) {
...
+
+    /**
+     * If Elasticsearch uses SSL with mutual authentication (via shield),
+     * provide the password to open the client keystore.
+     *
+     * @param keystorePassword the password of the client keystore.
+     * @return the {@link ConnectionConfiguration} object with keystore password set.
+     */
+    public ConnectionConfiguration withKeystorePassword(String keystorePassword) {
+        checkArgument(keystorePassword != null, "ConnectionConfiguration.create()"
```
Comment (jbonofre) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java]: Theoretically, the keystore password can be empty.

**Comment by jbonofre** [comment]
Rebased and addressed the comments.

**Comment by jbonofre** [comment]
retest this please

**Comment by echauchot** [comment]
LGTM, we need to update ITest (https://issues.apache.org/jira/browse/BEAM-2737) soon after :)

**Comment by jbonofre** [comment]
Thanks, I will merge and sync with @ssisk and @chamikaramj about the itests.

**Pull request merged**